src/ray/raylet/lineage_cache.cc

```diff
@@ -92,27 +92,49 @@ boost::optional<LineageEntry &> Lineage::GetEntryMutable(const TaskID &task_id)
 bool Lineage::SetEntry(const Task &task, GcsStatus status) {
   // Get the status of the current entry at the key.
   auto task_id = task.GetTaskSpecification().TaskId();
-  auto current_entry = GetEntryMutable(task_id);
-  if (current_entry) {
-    if (current_entry->SetStatus(status)) {
+  auto it = entries_.find(task_id);
+  bool updated = false;
+  if (it != entries_.end()) {
+    if (it->second.SetStatus(status)) {
+      // The task's spec may have changed, so remove its old dependencies.
+      // TODO(swang): This could be inefficient for tasks that have lots of
+      // dependencies and/or large specs. A flag could be passed in for tasks
```

Comment:
Why is it an issue if the task spec is large?

Reply:
Just that you have to copy the task spec over when updating the task entry. Updating the dependencies is probably going to be worse in most cases. I made this slightly better in the last commit.